### DISCUSSION TITLE
NBVM not able to find existing VM on IPv4 Address

### DISCUSSION THREAD

**Discussion by tkn2023** [bb-Ricardo/netbox-sync#381] [General]
Hi,
I recently discovered this masterpiece and tried it on my test-system and it worked. 
However I did not have any VM's in the test-system that exist in vcenter so I did not encounter this issue.
In the live enviroment I ran the script and it created every vm even tho the vm already existed and a primary ip was assigned.
In the logs I even encountered the issue:
```
DEBUG2: No match found. Trying to find virtual machine based on primary IP addresses
DEBUG: No existing virtual machine object for vmname. Creating a new virtual machine.
DEBUG: Current interface 'ethernet0 (vmname)' for IP '172.26.1.34/24' and this one 'vNIC 1 (new_vlan) (vmname)' are both enabled.
```
Which indicates for me that the primary IP address search is not working correctly for NBVM.
Any idea what caused this? New Netbox version (this is 3.7.4)?
Every bit helps. So feel free to ask anything you want or suggest anything and I will try this.

**Comment by tkn2023** [comment]
I was able to get this to work:
`def _matches_device_primary_ip(device_primary_ip, ip_needle):

            ip = None
            log.debug3(f"{device_primary_ip} and {ip_needle}")
            if device_primary_ip is not None and ip_needle is not None:
                if isinstance(device_primary_ip, dict):
                    ip = device_primary_ip.get("address")  # Access the 'address' key directly
                elif isinstance(device_primary_ip, str):  
                    ip = device_primary_ip.split("/")[0]  # Use it directly if it's a string
                #log.debug(f"Compare {ip} with {ip_needle}")
                if ip is not None and ip == ip_needle:  # Compare directly without splitting
                    return True
            return False

        if object_type not in [NBDevice, NBVM]:
            raise ValueError(f"Object must be a '{NBVM.name}' or '{NBDevice.name}'.")

        if primary_ip4 is None and primary_ip6 is None:
            return

        if primary_ip4 is not None:
            primary_ip4 = str(primary_ip4).split("/")[0]
            #log.debug3(f"Using {primary_ip4} as IP")
        if primary_ip6 is not None:
            primary_ip6 = str(primary_ip6).split("/")[0]

        for device in self.inventory.get_all_items(object_type):
            #debugip_address = grab(device, "data.primary_ip.address")
            #log.debug3(f"debug IP grab {debugip_address} ip4 {primary_ip4}")
            if _matches_device_primary_ip(grab(device, "data.primary_ip.address"), primary_ip4) is True:
                log.debug2(f"Found existing host '{device.get_display_name()}' "
                           f"based on the primary IPv4 '{primary_ip4}'")
                return device

            if _matches_device_primary_ip(grab(device, "data.primary_ip6"), primary_ip6) is True:
                log.debug2(f"Found existing host '{device.get_display_name()}' "
                           f"based on the primary IPv6 '{primary_ip6}'")
                return device`

**Comment by bb-Ricardo** [comment]
Hi,

Glad you found a solution. Would you mind creating a Pull Request?
